numpy/_core/src/multiarray/ctors.c

```diff
@@ -1839,7 +1839,7 @@ PyArray_CheckFromAny_int(PyObject *op, PyArray_Descr *in_descr,
         else if (in_descr && !PyArray_ISNBO(in_descr->byteorder)) {
             PyArray_DESCR_REPLACE(in_descr);
         }
-        if (in_descr && in_descr->byteorder != NPY_IGNORE) {
+        if (in_descr && in_descr->byteorder != NPY_IGNORE && in_descr->byteorder != NPY_NATIVE) {
```

Comment:
This looks strange. If  `in_descr->byteorder != NPY_NATIVE` is false, then how does setting it to `NPY_NATIVE` cause problems if that is already the value?

Reply:
I agree it's weird but this is the most straightforward fix without a bigger refactor for the way `PyArray_DESCR_REPLACE` works.

See https://github.com/numpy/numpy/issues/28143#issuecomment-2584266824 for more context about why it's generally worthwhile to fix TSAN races, even if they appear benign.

Reply:
I  guess the deeper question is, why that is a problem for this case, but not for the other cases. You suggest the reason lies in PyArray_DESCR_REPLACE. My sense is that the bigger refactor would be preferable, even if more work.  At the least, a comment would be helpful so that we don't overlook the need for a better fix down the road.

Reply:
> I guess the deeper question is, why that is a problem for this case, but not for the other cases.

It's a problem for the built-in dtypes, because this is overwriting data that we already set when we import numpy.

For a user dtype, it's less likely that the input descriptor will effectively be a global variable. At least that's my understanding.

I'll let @seberg weigh in before working on this more.

Reply:
The problem with `PyArray_DESCR_REPLACE` is for new-style/user dtypes.  As far as I understand tsan complains because the byte-order is reset to the identical value even when `PyArray_DESCR_REPLACE` was not called.

Maybe it would be more readable to just duplicate the:
```
    in_descr->byteorder = NPY_NATIVE;
```
in *both* branches just above.

For the sake of new-style dtypes, I suppose could replace this with an `ensure_canonical`, if @ngoldbaum agrees, I think I did that in some other places.  (I am not sure if `PyArray_DESCR_REPLACE` preserves metadata right now, nor `ensure_canonical` on our dtypes FWIW, but that would be the only change and it doesn't worry my.)

Reply:
I added a new `PyArray_DESCR_REPLACE_CANONICAL` macro to the internal `dtypemeta.h` header. Seemed generic enough to put it in a header instead of just replacing the one usage. It looks like there's one other spot it could be used in `descriptor.c` too:

https://github.com/numpy/numpy/blob/bbf4836a82d255c2a2161d277abb4b6f5455ca6d/numpy/_core/src/multiarray/descriptor.c#L1954

It looks like all the other internal uses are for dealing with unsized dtypes.